src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKStyle.java

```diff
@@ -182,6 +182,30 @@ else if (type == GTKColorType.WHITE) {
             }
         }
 
+        if (id == Region.LABEL && type == ColorType.FOREGROUND
+                && (state & SynthConstants.ENABLED) != 0) {
+            return getLAFDefinedColor(context, state, type, "Label.foreground");
+        }
+
+        if (id == Region.CHECK_BOX && type == ColorType.FOREGROUND
+                && (state & SynthConstants.DISABLED) != 0) {
+            return getLAFDefinedColor(context, state, type, "CheckBox.disabledText");
+        }
+
+        if (id == Region.RADIO_BUTTON && type == ColorType.FOREGROUND
+                && (state & SynthConstants.DISABLED) != 0) {
+            return getLAFDefinedColor(context, state, type, "RadioButton.disabledText");
+        }
+
+        return getStyleSpecificColor(context, state, type);
+    }
+
+    private Color getLAFDefinedColor(SynthContext context, int state,
+                                     ColorType type, String propertyName) {
+        Color c = (Color) get(context, propertyName);
```

Comment:
I guess these condition checks should also be done inside the method since you are passing the values..
Also, the enabled state are still not handled..If the existing testcase does not catch those, maybe we can add subtests to it..

Reply:
> Also, the enabled state are still not handled

The enabled state is not handled in Metal LAF also, tried adding the enabledText property by setting ui property --
`UIManager.getDefaults().put("CheckBox.enabledText", checkboxColor);`

but the checkbox/radiobutton is painted with default black color.
Should we handle the change in metal LAF as well?

Reply:
Backtracking a bit, regarding `Label.foreground` why it needs to be done here?

Other L&F, it is added in corresponding LookAndFeel class
for example, in [Metal](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/plaf/metal/MetalLookAndFeel.java#L1001)
and in [Windows](https://github.com/openjdk/jdk/blob/master/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/WindowsLookAndFeel.java#L880)
etc
so I think it is needed to be added in GTKLookAndFeel class

and additionally there is setting of `Label.foreground` and other Label. property in BasicLabelUI.java [installDefaults](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicLabelUI.java#L368)
but we dont do it for Nimbus or GTK or Synth so I guess we need to call `super.installDefaults(c)` in SynthLabelUI class since there is no GTKLabelUI class
or 
maybe do
`LookAndFeel.installColorsAndFont(c, "Label.background", "Label.foreground", "Label.font");` alone to avoid other properties being set and to not fallback

Please check if it will work
and if it does, then we need to employ similar for checbox/radiobutton/togglebutton

Reply:
I will check and update.

Reply:
> Backtracking a bit, regarding `Label.foreground` why it needs to be done here?
> 
> Other L&F, it is added in corresponding LookAndFeel class for example, in [Metal](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/plaf/metal/MetalLookAndFeel.java#L1001) and in [Windows](https://github.com/openjdk/jdk/blob/master/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/WindowsLookAndFeel.java#L880) etc so I think it is needed to be added in GTKLookAndFeel class

IN GTKLookAndFeel, the color is returned specific to the style and state from native side . So, even if it is not added the default color for enabled, disabled or any other state is returned by native. Whereas In Metal the color is set by using UIManager define color.

> and additionally there is setting of Label.foreground and other Label. property in BasicLabelUI.java [installDefaults](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicLabelUI.java#L368)
but we dont do it for Nimbus or GTK or Synth so I guess we need to call super.installDefaults(c) in SynthLabelUI class since there is no GTKLabelUI class
or
maybe do
LookAndFeel.installColorsAndFont(c, "Label.background", "Label.foreground", "Label.font"); alone to avoid other properties being set and to not fallback

This properties are set via SynthStyle `installDefaults` method where the foreground, background are set by getting the value from `getColorForState` method. So, calling super.installDefaults(c) in SynthLabelUI will not make any difference.

I checked for JLabel properties set in enabled and disabled state. 

```
In Nimbus LAF
            UIManager.getDefaults().put("Label[Enabled].textForeground", labelColor);
            UIManager.getDefaults().put("Label[Disabled].textForeground", labelDisabledColor);

Other LAF
            UIManager.getDefaults().put("Label.foreground", labelColor);
            UIManager.getDefaults().put("Label.disabledForeground", labelDisabledColor);
```

It works as expected for Metal and GTK. For Nimbus LAF, disabled color is not painted correctly. In stead the default color specified in skin.laf is rendered.

Reply:
> It works as expected for Metal and GTK. For Nimbus LAF, disabled color is not painted correctly. In stead the default color specified in skin.laf is rendered.

For HTML text, disabled state doesn't render in LAF defined color for Metal and GTK as well.